R/keyword_quote_linter.R

```diff
@@ -0,0 +1,162 @@
+#' Block unnecessary quoting in calls
+#'
+#' Any valid symbol can be used as a keyword argument to an R function call.
+#'   Sometimes, it is necessary to quote (or backtick) an argument that is
+#'   not an otherwise valid symbol (e.g. creating a vector whose names have
+#'   spaces); besides this edge case, quoting should not be done.
+#'
+#' The most common source of violation for this is creating named vectors,
+#'   lists, or data.frame-alikes, but it can be observed in other calls as well.
+#'
+#' Similar reasoning applies to extractions with `$` or `@`.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = 'data.frame("a" = 1)',,
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = "data.frame(`a` = 1)",
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = 'my_list$"key"',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = 's4obj@"key"',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "data.frame(`a b` = 1)",
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' lint(
+#'   text = 'my_list$`a b`',
+#'   linters = keyword_quote_linter()
+#' )
+#'
+#' @evalRd rd_tags("keyword_quote_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+# TODO(michaelchirico): offer a stricter version of this that
+#   requires backticks to be used for non-syntactic names (i.e., not quotes).
+#   Here are the relevant xpaths:
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/SYMBOL_SUB[starts-with(text(), '`')]
+#   //expr[expr[SYMBOL_FUNCTION_CALL]]/STR_CONST[{is_quoted(text())}]
+keyword_quote_linter <- function() {
+  # NB: xml2 uses xpath 1.0 which doesn't support matches() for regex, so we
+  #   have to jump out of xpath to complete this lint.
+  # It's also a bit tough to get the escaping through R and then xpath to
+  #   work as intended, hence the rather verbose declaration here.
+  quote_cond <- xp_or(
+    "starts-with(text(), '\"')",
+    "starts-with(text(), '`')",
+    'starts-with(text(), "\'")'
+  )
+  # SYMBOL_SUB for backticks, STR_CONST for quoted names
+  call_arg_xpath <- glue("
+  //SYMBOL_FUNCTION_CALL
+    /parent::expr
+    /parent::expr
+    /*[(self::SYMBOL_SUB or self::STR_CONST) and {quote_cond}]
+  ")
+
+  # also exclude $ or @, which are handled below
+  assignment_xpath <- "
+  (//EQ_ASSIGN | //LEFT_ASSIGN[text() != ':='])
```

Comment:
Is it worth supporting RIGHT_assign, too?

Reply:
Sure, though it's not as simple as just adding `//RIGHT_ASSIGN` since we need to change the sibling direction. Just a few lines.